modules/network/keeper/msg_server.go

```diff
@@ -0,0 +1,204 @@
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+// Attest handles MsgAttest
+func (k msgServer) Attest(goCtx context.Context, msg *types.MsgAttest) (*types.MsgAttestResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if !k.IsCheckpointHeight(ctx, msg.Height) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "height %d is not a checkpoint", msg.Height)
+	}
+	has, err := k.IsInAttesterSet(ctx, msg.Validator)
+	if err != nil {
+		return nil, errors.Wrapf(err, "in attester set")
+	}
+	if !has {
+		return nil, errors.Wrapf(sdkerrors.ErrUnauthorized, "validator %s not in attester set", msg.Validator)
+	}
+
+	index, found := k.GetValidatorIndex(ctx, msg.Validator)
+	if !found {
+		return nil, errors.Wrapf(sdkerrors.ErrNotFound, "validator index not found for %s", msg.Validator)
+	}
+
+	// todo (Alex): we need to set a limit to not have validators attest old blocks. Also make sure that this relates with
+	// the retention period for pruning
+	bitmap, _ := k.GetAttestationBitmap(ctx, msg.Height)
+	if bitmap == nil {
+		validators, err := k.stakingKeeper.GetLastValidators(ctx)
+		if err != nil {
+			return nil, err
+		}
+		numValidators := 0
+		for _, v := range validators {
+			if v.IsBonded() {
+				numValidators++
+			}
+		}
+		bitmap = k.bitmapHelper.NewBitmap(numValidators)
+	}
+
+	if k.bitmapHelper.IsSet(bitmap, int(index)) {
+		return nil, errors.Wrapf(sdkerrors.ErrInvalidRequest, "validator %s already attested for height %d", msg.Validator, msg.Height)
+	}
+
+	// TODO: Verify the vote signature here once we implement vote parsing
+
+	// Set the bit
+	k.bitmapHelper.SetBit(bitmap, int(index))
+	if err := k.SetAttestationBitmap(ctx, msg.Height, bitmap); err != nil {
+		return nil, errors.Wrap(err, "failed to set attestation bitmap")
+	}
+
+	// Store signature using the new collection method
+	if err := k.SetSignature(ctx, msg.Height, msg.Validator, msg.Vote); err != nil {
+		return nil, errors.Wrap(err, "failed to store signature")
+	}
+
+	epoch := k.GetCurrentEpoch(ctx)
+	epochBitmap := k.GetEpochBitmap(ctx, epoch)
+	if epochBitmap == nil {
+		validators, err := k.stakingKeeper.GetLastValidators(ctx)
+		if err != nil {
+			return nil, err
+		}
+		numValidators := 0
+		for _, v := range validators {
+			if v.IsBonded() {
+				numValidators++
+			}
+		}
+		epochBitmap = k.bitmapHelper.NewBitmap(numValidators)
+	}
+	k.bitmapHelper.SetBit(epochBitmap, int(index))
+	if err := k.SetEpochBitmap(ctx, epoch, epochBitmap); err != nil {
+		return nil, errors.Wrap(err, "failed to set epoch bitmap")
+	}
+
+	// Emit event
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.TypeMsgAttest,
+			sdk.NewAttribute("validator", msg.Validator),
+			sdk.NewAttribute("height", math.NewInt(msg.Height).String()),
+		),
+	)
+
+	return &types.MsgAttestResponse{}, nil
+}
+
+// JoinAttesterSet handles MsgJoinAttesterSet
```

Comment:
I think this should be gov gated. I am not seeing that the attesters vote are weighted by stake, so it would be very easy to spam this and slow down the chain, as well as prohibit the attester set to reach consensus.

Reply:
EDIT: i see vp is calculated in the endblocker, so my concerns about spamming is gone.